src/Lecture2.hs

```diff
@@ -70,7 +80,14 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt index l = rem 0 l []
+    where
+      rem :: Int -> [a] -> [a] -> (Maybe a, [a])
+      rem _ [] acc     = (Nothing, acc)
+      rem i (x:xs) acc
+          | i == index = (Just x, acc ++ xs)
+          | otherwise  = rem (i + 1) xs (acc ++ [x])
```

Comment:
This looks correct 👍🏻 

However, this function hangs when given a negative index and an infinite list ♾️

Reply:
Ok I understand the problem with negative number but I can't figure out why it hangs on infinite list

Reply:
You will stop only if you reach an empty list or on the `i == index`.
Since the list is infinite, you won't reach an infinite list.

Well, you're using `Int`, so, eventually, due to integer overflow, you'll reach the index. But it's a long road!

The simplest solution is to return `Nothing` and the list itself when the given index is negative. Bonus points for checking index negativity only once in the beginning and not on every recursive call 🙂